hail/python/hail/table.py

```diff
@@ -1497,12 +1497,37 @@ def joiner(left):
                                               f"  Table key:         {', '.join(str(t) for t in self.key.dtype.values())}\n"
                                               f"  Index Expressions: {', '.join(str(e.dtype) for e in exprs)}")
 
-                if is_subset_row_key or is_interval:
-                    key = None
+                if not (is_subset_row_key or is_interval):
+                    # foreign-key join
+                    foreign_key_annotates = {Env.get_uid(): e for e in exprs}
+
+                    # contains original key and join key
+                    join_table = src.select_rows(**foreign_key_annotates).rows()
+
+                    join_table = join_table.key_by(*foreign_key_annotates)
+
+                    value_uid = Env.get_uid()
+                    join_table = join_table.annotate(**{value_uid: right.index(join_table.key)})
+
+                    #  FIXME: TableZipUnchecked would make this faster
```

Comment:
Is this a slow operation, and if so, would it make sense to address it now?

Reply:
This is the implementation of a foreign-key row join, and has to do two key_by operations. It may be possible to shuffle less data by using row index and zipping, which would require the TableZipUnchecked node which hasn't yet been added -- it's added in another open PR which has some limitations that need to be fixed before it can go in.


I'll just remove the FIXME for now - I'm not even sure that using TableZipUnchecked would be an overall win because it could constrain optimization in ways I haven't fully thought through.

Reply:
Sounds good! Thanks for the explanation